Refactor how COI configures Deltachat behavior

mneumann opened the pull request:
* Store all information that influence Deltachat's behavior in case COI
  is present in struct `CoiDeltachatMode`.

* Access `CoiDeltachatMode` via a Mutex. No fancy AtomicBool anymore.
  Performance doesn't really matter here.

* Refactor some commonly used code into `mvbox_use_network()`. We might
  want to override the database settings in case of COI.

* CoiConfig - hardcode imap delimiter to '/'.  AFAIK, there is no way to
  determine which IMAP delimiter a server uses.  For COI we can assume a
  Dovecot server. But this is a more general problem and has to be
  solved on a higer level.

mneumann committed:
Refactor how COI configures Deltachat behavior

* Store all information that influence Deltachat's behavior in case COI
  is present in struct `CoiDeltachatMode`.

* Access `CoiDeltachatMode` via a Mutex. No fancy AtomicBool anymore.
  Performance doesn't really matter here.

* Refactor some commonly used code into `mvbox_use_network()`. We might
  want to override the database settings in case of COI.

* CoiConfig - hardcode imap delimiter to '/'.  AFAIK, there is no way to
  determine which IMAP delimiter a server uses.  For COI we can assume a
  Dovecot server. But this is a more general problem and has to be
  solved on a higer level.

VP- commented on src/job.rs:
```diff
@@ -516,30 +515,28 @@ pub fn perform_imap_idle(context: &Context) {
         .get_config_int(context, "mvbox_watch")
-        .unwrap_or_else(|| 1);
+        .map(|value| value == 1)
+        .unwrap_or(true)
```
How about `.unwrap_or(1) == 1`?

mneumann replied:
yes! best would be to add a ```get_config_bool()``` method because we have this kind of code all over the place.

mneumann replied:
The `imap_delimiter` part of `ImapConfig` is not updated with the value from config. Its default of "." is "wrong" for Dovecot. Also, this shouldn't be something the user can set. I created a separate ticket for that issue.

VP- commented on src/job.rs:
```diff
@@ -516,30 +515,28 @@ pub fn perform_imap_idle(context: &Context) {
         .write()
         .unwrap()
-        .fetch(context, use_network == 1);
+        .fetch(context, mvbox_use_network(context));
```
You're holding the write lock while performing an SQL query.

mneumann replied:
not good

VP- commented on src/job.rs:
```diff
@@ -516,30 +515,28 @@ pub fn perform_imap_idle(context: &Context) {
         .read()
         .unwrap()
-        .idle(context, use_network == 1);
+        .idle(context, mvbox_use_network(context));
```
Same as above.

VP- commented on src/coi/deltachat_mode.rs:
```diff
@@ -0,0 +1,45 @@
+    pub server_side_move_enabled: bool,
+    pub inbox_folder_override: Option<String>,
+    pub mvbox_folder_override: Option<String>,
+}
```
Do we really need to have the separation between COI and DeltaChat be visible outside (i.e. make everything conditional on `server_side_move_enabled`? Just having an `inbox_folder` and an optional `mvbox_folder` would probably remove some `if`s throughout the code base. Only the code writing to "configured_mvbox_folder" might need to be changed to update the mode here.

mneumann replied:
The reason why `server_side_move_enabled` is there is that we have to disable moving messages in `dc_do_heuristics_moves` for instance if COI is active. I don't want to overload the meaning of the "overrides" too much. Overriding the inbox path should not necessarily mean that Coi is enabled. Actually, without the unit tests, we could get rid of `inbox_folder_override` and use "INBOX" all the time.

VP- replied:
My question was about the dependency between the flag and the folder names, not about the existence of the flag. The idea was instead of having a folder name and an override for it, to just have one variables for the final folder name (one for each thread). So that `dc_is_*` called by `do_heuristics_moves` just compare with the stored string instead of checking the flag and reading the config each time.

mneumann replied:
You mean a configuration for each thread, which folder name to use, unrelated to Coi.
Yes, this would have been a better approach in the first place.

VP- replied:
Depends on what you mean by "configuration". In DCC this usually means persistent entries in the SQLite database. With "variable" I mean a variable in RAM. Like what the `*_folder_override` variables already are, except without the `_override` suffix, and reading the configured mvbox at most once, if client-side filtering is enabled (as opposed to every time `dc_is_mvbox` is called).

The configuration in the DB would stay the same, i.e. store the folder which is used for client-side filtering.

VP- reviewed (commented):
The delimiter is part of `ImapConfig` and can be configured by the user if the default doesn't work.

mneumann committed:
Address some of Viktors comments

* Get rid of `server_side_move_enabled`. CoiDeltachatMode is now an
  enum. The overrides are only available in the `Enabled` case.

* Get rid of `impl Default for CoiDeltachatMode`. Defaults are not the
  Rust way. Defaults kind of hide uninitialized state.

* Don't access the database while holding a write lock!

* Add `Sql#get_config_bool`. This treats every integer value != 0 as
  true whereas the original code treated only 1 as true (which is very
  counter intunitive for "C" code).
